src/sknnr_spatial/utils/estimator.py

```diff
@@ -25,11 +29,16 @@ def __dict__(self):
         return self._wrapped.__dict__
 
 
-def check_wrapper_implements(func: Callable) -> Callable:
+GenericWrapper = TypeVar("GenericWrapper", bound=AttrWrapper)
+
+
+def check_wrapper_implements(
+    func: Callable[Concatenate[GenericWrapper, P], RT],
+) -> Callable[Concatenate[GenericWrapper, P], RT]:
```

Comment:
> The one thing I don't understand is why you have to create the separate GenericWrapper type variable rather than just using AttrWrapper directly.

Using `AttrWrapper` as the parameter here raises a mypy error for all of the decorated `ImageEstimator` methods:

```raw
Argument 1 to "check_wrapper_implements" has incompatible type "Callable[[ImageEstimator[EstimatorType], Any, Any, KwArg(Any)], ImageEstimator[EstimatorType]]"; expected "Callable[[AttrWrapper[Any], Any, Any, KwArg(Any)], ImageEstimator[EstimatorType]]"  [arg-type]
```

The error seems to boil down to expecting an `AttrWrapper` but getting an `ImageEstimator`. I had assumed that `AttrWrapper` would cover subclasses including `ImageEstimator`, but after reading some [StackOverflow answers](https://stackoverflow.com/a/59937840/17707273) and skimming [PEP 484](https://peps.python.org/pep-0484/#type-variables-with-an-upper-bound), I think that's not the case. From PEP 484:

> Consider a class Employee with a subclass Manager. Now suppose we have a function with an argument annotated with List[Employee]. Should we be allowed to call this function with a variable of type List[Manager] as its argument? Many people would answer “yes, of course” without even considering the consequences. But unless we know more about the function, a type checker should reject such a call: the function might append an Employee instance to the list, which would violate the variable’s type in the caller.

Using a bound `TypeVar` avoids that issue by explicitly telling `mypy` that we can swap any `AttrWrapper` subclass in without issues. Also from PEP 484:

> A type variable may specify an upper bound using bound=type (note: type itself cannot be parameterized by type variables). This means that an actual type substituted (explicitly or implicitly) for the type variable must be a subtype of the boundary type.

Reply:
Excellent, thanks for the detailed answer here.  I'm sure I'll still get it wrong, but this explanation did help.